library/core/src/intrinsics.rs

```diff
@@ -1207,29 +1207,32 @@ extern "rust-intrinsic" {
 
     /// Reinterprets the bits of a value of one type as another type.
     ///
-    /// Both types must have the same size. Neither the original, nor the result,
-    /// may be an [invalid value](../../nomicon/what-unsafe-does.html).
+    /// Both types must have the same size. Compilation will fail if this is not guaranteed.
     ///
     /// `transmute` is semantically equivalent to a bitwise move of one type
     /// into another. It copies the bits from the source value into the
-    /// destination value, then forgets the original. It's equivalent to C's
-    /// `memcpy` under the hood, just like `transmute_copy`.
+    /// destination value, then forgets the original. Note that source and destination
+    /// are passed by-value, which means if `T` or `U` contains padding, that padding
+    /// might *not* be preserved by `transmute`.
+    ///
+    /// Both the argument and the result must be [valid](../../nomicon/what-unsafe-does.html) at
+    /// their given type. Violating this condition leads to [undefined behavior][ub]. The compiler
+    /// will generate code *assuming that you, the programmer, ensure that there will never be
+    /// undefined behavior*. It is therefore your responsibility to guarantee that every value
+    /// passed to `transmute` is valid at both types `T` and `U`. Failing to uphold this condition
+    /// may lead to unexpected and unstable compilation results. This makes `transmute` **incredibly
+    /// unsafe**. `transmute` should be the absolute last resort.
+    ///
+    /// Transmuting pointers to integers in a `const` context is [undefined behavior][ub].
```

Comment:
Is it worth mentioning that it's discouraged in non-const contexts? We mention further down that "integer-to-pointer `transmute`" is (somehow) "problematic" below, in documentation which applies to non-const. I think that's enough justification to note it here as well if you think it's warranted.

That said, I don't think we should be too specific about it (strict provenance is still an experiment, nominally, and provenance models that allow this do exist IIUC), but even in a hypothetical world where provenance or lack there-of can't cause you any problems, using `as` for ptr->int/int->ptr would still be preferable as it's safe, unlike transmute.

Don't feel the need to note it though -- we can always just wait until things are settled enough that we don't need to be vague about it.

Reply:
I have added a parenthetical, please have a look. I want to make sure I am not wording this stronger than what is actually justified.

Reply:
That looks great.